eth/catalyst/simulated_beacon.go

```diff
@@ -0,0 +1,231 @@
+	for {
+		select {
+		case withdrawal := <-w.pending:
+			withdrawals = append(withdrawals, withdrawal)
+			if len(withdrawals) == maxCount {
+				break
+			}
+		default:
+			return withdrawals
+		}
+	}
+}
+
+type SimulatedBeacon struct {
+	shutdownCh  chan struct{}
+	eth         *eth.Ethereum
+	period      uint64
+	withdrawals withdrawalQueue
+
+	feeRecipient     common.Address
+	feeRecipientLock sync.Mutex // lock gates concurrent access to the feeRecipient
+
+	engineAPI          *ConsensusAPI
+	curForkchoiceState engine.ForkchoiceStateV1
+	lastBlockTime      uint64
+}
+
+func NewSimulatedBeacon(period uint64, eth *eth.Ethereum) (*SimulatedBeacon, error) {
+	chainConfig := eth.APIBackend.ChainConfig()
+	if !chainConfig.IsDevMode {
+		return nil, errors.New("incompatible pre-existing chain configuration")
+	}
+	block := eth.BlockChain().CurrentBlock()
+	current := engine.ForkchoiceStateV1{
+		HeadBlockHash:      block.Hash(),
+		SafeBlockHash:      block.Hash(),
+		FinalizedBlockHash: block.Hash(),
+	}
+	engineAPI := NewConsensusAPI(eth)
+
+	// if genesis block, send forkchoiceUpdated to trigger transition to PoS
+	if block.Number.Sign() == 0 {
+		if _, err := engineAPI.ForkchoiceUpdatedV2(current, nil); err != nil {
+			return nil, err
+		}
+	}
+	return &SimulatedBeacon{
+		eth:                eth,
+		period:             period,
+		shutdownCh:         make(chan struct{}),
+		engineAPI:          engineAPI,
+		lastBlockTime:      block.Time,
+		curForkchoiceState: current,
+		withdrawals:        withdrawalQueue{make(chan *types.Withdrawal, 20)},
+	}, nil
+}
+
+func (c *SimulatedBeacon) setFeeRecipient(feeRecipient common.Address) {
+	c.feeRecipientLock.Lock()
+	c.feeRecipient = feeRecipient
+	c.feeRecipientLock.Unlock()
+}
+
+// Start invokes the SimulatedBeacon life-cycle function in a goroutine.
+func (c *SimulatedBeacon) Start() error {
+	if c.period == 0 {
+		go c.loopOnDemand()
+	} else {
+		go c.loop()
+	}
+	return nil
+}
+
+// Stop halts the SimulatedBeacon service.
+func (c *SimulatedBeacon) Stop() error {
+	close(c.shutdownCh)
+	return nil
+}
+
+// sealBlock initiates payload building for a new block and creates a new block
+// with the completed payload.
+func (c *SimulatedBeacon) sealBlock(withdrawals []*types.Withdrawal) error {
+	tstamp := uint64(time.Now().Unix())
+	if tstamp <= c.lastBlockTime {
+		tstamp = c.lastBlockTime + 1
+	}
+	c.feeRecipientLock.Lock()
+	feeRecipient := c.feeRecipient
+	c.feeRecipientLock.Unlock()
+
+	fcResponse, err := c.engineAPI.ForkchoiceUpdatedV2(c.curForkchoiceState, &engine.PayloadAttributes{
+		Timestamp:             tstamp,
+		SuggestedFeeRecipient: feeRecipient,
+		Withdrawals:           withdrawals,
+	})
+	if err != nil {
+		return fmt.Errorf("error calling forkchoice update: %v", err)
+	}
+
+	envelope, err := c.engineAPI.getFullPayload(*fcResponse.PayloadID)
+	if err != nil {
+		return fmt.Errorf("error retrieving payload: %v", err)
+	}
+	payload := envelope.ExecutionPayload
```

Comment:
I haven't been following the developments lately. A while back, it was designed so that there was always a block in production, and new transactions would be added to it, and whenever we wanted to, we could collect it. Meaning that the sealer has a lot of time to seal it: if the time between blocks is 10 seconds, the prior design would have 10 seconds to work on each block. 

How is it now? It _looks_ to me like we are giving the miner zero time to work on the block, before we collect it. Am I missing something?

Reply:
With the new API, if there are transactions in the pool `getFullPayload` will hang until a full block is created so we no longer need delay logic within the `SimulatedBeacon`.